rotkehlchen/chain/ethereum/uniswap/uniswap.py

```diff
@@ -220,6 +244,224 @@ def _get_known_asset_price(
 
         return asset_price
 
+    def _get_trades(
+            self,
+            addresses: List[ChecksumEthAddress],
+            from_timestamp: Timestamp,
+            to_timestamp: Timestamp,
+            db_query_get_used_query_range: Callable,
+            db_query_update_used_query_range: Callable,
+            db_query_get_trades: Callable,
+            db_query_add_trades: Callable,
+            get_ammtrade_from_trade: Callable,
+    ) -> AddressTrades:
+        """Request via graph all trades for new addresses and the latest ones
+        for already existing addresses. Then the requested trade are stored in
+        DB and finally all DB trades are returned.
+
+        Requested trades are managed as <AMMTrade>, whilst trades read from DB
+        and insterted to DB are managed as <Trade> (with a custom identifer).
+        """
+        address_ammtrades: AddressTrades = {}
+        db_address_trades: DDAddressTrades = defaultdict(list)
+        new_addresses: List[ChecksumEthAddress] = []
+        existing_addresses: List[ChecksumEthAddress] = []
+        min_end_ts: Timestamp = to_timestamp
+
+        # Get addresses' last used query range for Uniswap trades
+        for address in addresses:
+            entry_name = f'{UNISWAP_TRADES_PREFIX}_{address}'
+            trades_range = db_query_get_used_query_range(name=entry_name)
+
+            if not trades_range:
+                new_addresses.append(address)
+            else:
+                existing_addresses.append(address)
+                min_end_ts = min(min_end_ts, trades_range[1])
+
+        # Request new addresses' trades
+        if new_addresses:
+            start_ts = Timestamp(0)
+            new_address_trades = self._get_trades_graph(
+                addresses=new_addresses,
+                graph_query=self.graph.query,  # type: ignore
```

Comment:
If you were using the graph object as `self.graph` in `_get_trades_graph` you would not be having this type problem

Reply:
from `mypy`:
```shell
rotkehlchen/chain/ethereum/uniswap/uniswap.py:366: error: Item "None" of "Optional[Graph]" has no attribute "query"
```
I've been putting the type ignore in balances and here because in this module `self.graph` can be `None`.

Reply:
In fact you put the same `# type: ignore` in the Compound module :P

Reply:
I think you may not understand what the type system is good for. The warning is correct then and you should not ignore it, especially without a comment explaining why you are ignoring. We don't use mypy because I am a masochist, but because it helps us find bugs. And that warning is correct and shows you have a bug.

Indeed initializing the graph object may fail and then it's none. And if you get in this funcion with `self.graph == None` what will happen? An exception will occur.

Do you check anywhere on the call chain that graph is not None? If not then just add a check in the start of the function and return a failure or raise an exception. Then mypy will not complain because you will have indeed verified that `self.graph` is not None.

Reply:
> In fact you put the same ...

I do in the subfunctions, because where I call them from I verify that self.graph is not None.

```
        if self.graph is None:  # could not initialize graph
            return {}
```

https://github.com/rotki/rotki/blob/develop/rotkehlchen/chain/ethereum/compound.py#L657

Reply:
I see, my bad on the historical endpoint because definitely if `self.graph` is None I must return.
But for balances, if I ever refactor the code back to `self.graph.query` (instead of the fake injection) I'm pretty sure I'll need the ignore comment, unless I check in each function that is not None. I will take it into account for history, thanks!

Reply:
> But for balances, if I ever refactor the code back to self.graph.query

We discussed in the morning you would not do the fake injection, no?

> I'll need the ignore comment, unless I check in each function that is not None. I will take it into account for history, thanks!

If the calling function already does a check, then it's fine to put in the ignore comment. Perhaps just leave a comment why you ignore. Something like

```python
self.graph.foo()  # type: ignore - caller already checks
```

Reply:
Yes, as agreed this morning no more fake injections, but they are still present in the get balances endpoint. For getting rid of them I'll need to refactor the tests. For that reason I will leave it as it is for now.

Reply:
Hmmm I really dislike this coding pattern. I understand the time pressure point so okay but make an issue to remove it as soon as possible so we don't forget it

Reply:
Marking the comment as resolved and pointing to an issue: https://github.com/rotki/rotki/issues/1727